File: advent-of-code/bin/day01/day-01.js
#!/usr/bin/env node

var fs = require( 'fs' );

var santa = {
  floor: 0,
  step: 0,
  entered_basement: 0,

  goUp: function(){
    this.step++;
    this.floor++;
  },

  goDown: function(){
    this.step++;
    this.floor--;
    if ( this.floor < 0 && ! this.entered_basement )
      this.entered_basement = this.step;
  }
};


fs.readFile( __dirname + '/day-01.txt', function( err, data ) {
  if ( err ) {
    throw err;
  }
  s = data.toString();
  var l = s.length;
  var i;
  var c;
  for ( i = 0 ; i < l ; i++ ) {
    c = s.charAt(i);
    if      ( c == '(' ) santa.goUp()
    else if ( c == ')' ) santa.goDown()
  }
  console.log( "Santa ends up on floor", santa.floor );
  console.log( "Santa first enters the basement on step ", santa.entered_basement );
});

File: advent-of-code/src/day01/day-01.py
#!/usr/bin/env python

with open('day-01.txt','r') as input_file:
  data = input_file.read()

class Building:
  floor = 0
  _step = 0
  entered_basement = 0

  def step( self ):
    self._step += 1

  def goUp( self ):
    self.step()
    self.floor += 1

  def goDown( self ):
    self.step()
    self.floor -= 1
    if ( ( not self.entered_basement ) and ( self.floor < 0 ) ):
      self.entered_basement = self._step

santa = Building()
for instruction in data:
  if ( instruction == '(' ):
    santa.goUp()
  elif ( instruction == ')' ):
    santa.goDown()

print "Santa ended up on floor", santa.floor
print "Santa entered the basement on step ", santa.entered_basement

